test/client.nim

```diff
@@ -0,0 +1,65 @@
+import # nim libs
+  json, options, os, times, unittest
+
+import # vendor libs
+  chronos, confutils, json_serialization, sqlcipher, web3/conversions as web3_conversions
+
+import # nim-status libs
+  ../nim_status/[accounts, client, config, conversions, database, settings],
+  ./test_helpers
+
+procSuite "client":
+  asyncTest "client":
+
+    let dataDir = currentSourcePath.parentDir() / "build"
+    var config = StatusConfig.load(
+      @["--rootDataDir=" & dataDir]
+    )
+
+    let statusObj = init(config)
+
+    var account:Account = Account(
+      name: "Test",
+      loginTimestamp: 1,
+      identicon: "data:image/png;base64,something",
+      keycardPairing: "",
+      keyUid: "0x1234"
+    )
+
+    statusObj.saveAccount(account)
+    statusObj.updateAccountTimestamp(1, "0x1234")
+    let accounts = statusObj.getAccounts()
+    check:
+      $statusObj.config.rootDataDir == dataDir
+      accounts[0].keyUid == "0x1234"
+
+    let password = "qwerty"
+    statusObj.login(account.keyUid, password)
+    let settingsStr = """{
+      "address": "0x1122334455667788990011223344556677889900",
+      "chaos-mode": true,
+      "networks/current-network": "mainnet",
+      "dapps-address": "0x1122334455667788990011223344556677889900",
+      "eip1581-address": "0x1122334455667788990011223344556677889900",
+      "installation-id": "ABC-DEF-GHI",
+      "key-uid": "XYZ",
+      "latest-derived-path": 0,
+      "networks/networks": [{"id":"mainnet_rpc","etherscan-link":"https://etherscan.io/address/","name":"Mainnet with upstream RPC","config":{"NetworkId":1,"DataDir":"/ethereum/mainnet_rpc","UpstreamConfig":{"Enabled":true,"URL":"wss://mainnet.infura.io/ws/v3/7230123556ec4a8aac8d89ccd0dd74d7"}}}],
+      "name": "test",
+      "photo-path": "ABXYZC",
+      "preview-privacy?": false,
+      "public-key": "0x123",
+      "signing-phrase": "ABC DEF GHI",
+      "wallet-root-address": "0x1122334455667788990011223344556677889900"
+    }"""
+    let settingsObj = Json.decode(settingsStr, Settings, allowUnknownFields = true)
+
+    let nodeConfig = %* {"config": 1}
+    statusObj.createSettings(settingsObj, nodeConfig)
+    let dbSettings = statusObj.getSettings()
+    check:
+      dbSettings.keyUID == settingsObj.keyUID
+
+
+    statusObj.close()
+    removeDir(datadir)
```

Comment:
The `dataDir` is currently `test/build`, which is the same directory in which the compiled executable is located, so this line causes the executable to delete itself, which seems a bit strange, and may be problematic considering that `test/client.nim` is compiled into `test/build/test_all`, which runs more than one test that uses the `test/build` directory.

L14 could be changed to:
```nim
let dataDir = currentSourcePath.parentDir() / "build" / "data"
```

Or instead of `removeDir` you could use `removeFile` and delete the two `.db` files; see for example:
https://github.com/status-im/nim-status/blob/master/test/accounts.nim#L50

Reply:
Thanks, fixed it. Also added an extra `createDir` call, as otherwise `confutils.load()` will complain that `Error while processing the 'rootDataDir' parameter: Directory doesn't exist`.

Reply:
> Also added an extra `createDir` call, as otherwise `confutils.load()` will complain that `Error while processing the 'rootDataDir' parameter: Directory doesn't exist`.

@siphiuel that's an interesting problem, hmm... it wasn't happening when `dataDir` was defined as `currentSourcePath.parentDir() / "build"`, so I'm guessing `load()` has a problem when multiple directory-levels of the path don't exist.

Take a look at waku's config: https://github.com/status-im/nim-waku/blob/master/waku/v2/node/config.nim#L81-L84.

For their `dbpath` they use `string` instead of `InputDir` or `InputFile`. So maybe our config's `rootDataDir` and `accountsDbFileName` should be `string`. That way `load()` won't try to check that the paths already exist, and we already take care of making sure the directory exists (via `createDir`) in `initializeDB` so it should be okay.

Reply:
Right, indeed, that was caused by InputDir/InputFile doing some checking of their own. Converted to `string`.